File: Medium/JavaScript/2610. Convert an Array Into a 2D Array With Conditions.js
var findMatrix = function(nums) {
    let set = [];

    nums.forEach((n) => {
        let i = 0;
        while(i < set.length && set[i].has(n)){
            i++;
        }
        if(i === set.length)
            set.push(new Set([n]));
        else
            set[i].add(n);
    })

    for(let i in set){
        set[i] = Array.from(set[i]);
    }

    return set;
};

File: Medium/Java/2610. Convert an Array Into a 2D Array With Conditions.java
class Solution {
    public List<List<Integer>> findMatrix(int[] nums) {
        int n = nums.length;
        int[] freq = new int[n + 1];
        int max = 0;
        List<List<Integer>> list = new ArrayList<>();

        for(int num : nums){
            freq[num]++;
            max = Math.max(max, freq[num]);
        }

        for(int i = 0; i < max; i++){
            list.add(new ArrayList<>());
        }
        
        for(int i = 0; i <= n; i++){
            while(0 < freq[i]--){
                list.get(freq[i]).add(i);
            }
        }

        return list;
    }
}
